### PR TITLE
PGM integration

### PR DESCRIPTION
**Pull request by mrakitin**
This PR implements the PGM integration for the ARI example `00000003` or `00000004`.
We also added the GitHub Actions workflow to perform testing, based on the example from https://github.com/BNL-ATF/profile_atf.

### CODE REVIEW

**Comment by mrakitin** [comment]
## TODO for the next coding session:

### Expected angle values:
```py
In [2]: pgm.energy.put(250)

In [3]: pgm.read()
Out[3]: 
OrderedDict([('pgm__gratings',
              {'value': {'LowE': {'a0': 50,
                 'a1': 0.01868,
                 'a2': 1.95e-06,
                 'a3': 4e-09},
                'HighE': {'a0': 50,
                 'a1': 0.02986,
                 'a2': 2.87e-06,
                 'a3': 8e-09},
                'HighR': {'a0': 200,
                 'a1': 0.05743,
                 'a2': 6.38e-06,
                 'a3': 1.5e-08}},
               'timestamp': 1677109887.514582}),
             ('pgm_energy', {'value': 250.0, 'timestamp': 1677109887.5147781}),
             ('pgm_grating_name',
              {'value': 'HighR', 'timestamp': 1677109887.514876}),
             ('pgm_grated_harm_num',
              {'value': 1, 'timestamp': 1677109887.5149581}),
             ('pgm_pre_mirror_angle',
              {'value': 92.43679655196874, 'timestamp': 1677109887.515052}),
             ('pgm_grating_angle',
              {'value': 93.10512985252913, 'timestamp': 1677109887.515167}),
             ('pgm__r2', {'value': 11500, 'timestamp': 1677109887.515261}),
             ('pgm__r1', {'value': 32100, 'timestamp': 1677109887.515336}),
             ('pgm__m', {'value': 1, 'timestamp': 1677109887.5154262}),
             ('pgm__x_inc', {'value': 90, 'timestamp': 1677109887.5155141}),
             ('pgm__x_diff', {'value': 90, 'timestamp': 1677109887.515777}),
             ('pgm__b', {'value': 1, 'timestamp': 1677109887.515893}),
             ('pgm_cff',
              {'value': 1.7552544118691895, 'timestamp': 1677109887.516024}),
             ('pgm__a0', {'value': 200, 'timestamp': 1677109887.516124}),
             ('pgm__a1', {'value': 0.0579465, 'timestamp': 1677109887.516215}),
             ('pgm__a2', {'value': 6.3482e-06, 'timestamp': 1677109887.5163}),
             ('pgm__a3',
              {'value': 7.909e-10, 'timestamp': 1677109887.516386})])

```

### Calculated angles:
```py
In [1]: pgm.read()
Out[1]: 
OrderedDict([('pgm__gratings',
              {'value': {'LowE': {'a0': 50,
                 'a1': 0.01868,
                 'a2': 1.95e-06,
                 'a3': 4e-09},
                'HighE': {'a0': 50,
                 'a1': 0.02986,
                 'a2': 2.87e-06,
                 'a3': 8e-09},
                'HighR': {'a0': 200,
                 'a1': 0.05743,
                 'a2': 6.38e-06,
                 'a3': 1.5e-08}},
               'timestamp': 1677109887.514582}),
             ('pgm_energy', {'value': 250, 'timestamp': 1677109887.5147781}),
             ('pgm_grating_name',
              {'value': 'HighR', 'timestamp': 1677109887.514876}),
             ('pgm_grated_harm_num',
              {'value': 1, 'timestamp': 1677109887.5149581}),
             ('pgm_pre_mirror_angle',
              {'value': 91.94403078738465, 'timestamp': 1677109887.515052}),
             ('pgm_grating_angle',
              {'value': 91.79659034418579, 'timestamp': 1677109887.515167}),
             ('pgm__r2', {'value': 11500, 'timestamp': 1677109887.515261}),
             ('pgm__r1', {'value': 32100, 'timestamp': 1677109887.515336}),
             ('pgm__m', {'value': 1, 'timestamp': 1677109887.5154262}),
             ('pgm__x_inc', {'value': 90, 'timestamp': 1677109887.5155141}),
             ('pgm__x_diff', {'value': 90, 'timestamp': 1677109887.515777}),
             ('pgm__b', {'value': 1, 'timestamp': 1677109887.515893}),
             ('pgm_cff',
              {'value': 1.7552544118691895, 'timestamp': 1677109887.516024}),
             ('pgm__a0', {'value': 200, 'timestamp': 1677109887.516124}),
             ('pgm__a1', {'value': 0.0579465, 'timestamp': 1677109887.516215}),
             ('pgm__a2', {'value': 6.3482e-06, 'timestamp': 1677109887.5163}),
             ('pgm__a3',
              {'value': 7.909e-10, 'timestamp': 1677109887.516386})])
```

**Comment by mrakitin** [comment]
We observed an issue when the `.get()` output was not synchronized between the PGM object and the corresponding `objects[...]` element, and the `._sirepo_dict` was not returning the right value.

**Comment by mrakitin** [comment]
```py
In [23]: objects["grating"].cff.put(321)

In [24]: objects["grating"].cff.get()
Out[24]: 321

In [25]: pgm.cff._sirepo_dict
Out[25]: 
{'autocomputeVectors': 'vertical',
 'cff': 321,
 'computeParametersFrom': '2',
 'diffractionOrder': 1,
 'energyAvg': 250.0,
 'grazingAngle': 31.35641681558021,
 'grooveDensity0': 200,
 'grooveDensity1': 0.0579465,
 'grooveDensity2': 6.3482e-06,
 'grooveDensity3': 7.909e-10,
 'grooveDensity4': 0,
 'heightAmplification': 0.001,
 'heightProfileFile': 'ARIpdGRerr.txt',
 'horizontalOffset': 0,
 'id': 7,
 'normalVectorX': 0,
 'normalVectorY': -0.9985737995953669,
 'normalVectorZ': -0.05338882618743524,
 'nvx': 0,
 'nvy': 0.9995084278413885,
 'nvz': -0.031351278666681234,
 'orientation': 'y',
 'outframevx': 1.0,
 'outframevy': 0.0,
 'outoptvx': 0.0,
 'outoptvy': 0.08572697995704542,
 'outoptvz': 0.9963186663449823,
 'position': 32.1,
 'rollAngle': 0,
 'sagittalSize': 0.03,
 'tangentialSize': 0.2,
 'tangentialVectorX': 0,
 'tangentialVectorY': -0.05338882618743524,
 'title': 'Grating',
 'tvx': 0,
 'tvy': 0.031351278666681234,
 'type': 'grating',
 'verticalOffset': 0}

In [26]: pgm.cff.get()
Out[26]: 1.7552544118691895

In [27]: pgm.cff._readback
Out[27]: 1.7552544118691895

In [28]: pgm.cff._value
Out[28]: 1.7552544118691895

In [29]: 
```

File: startup/31-pgm-energy.py [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+
+    beta = -90 + b * (theta_gr - x_diff)
+    alpha = 180 + beta + b * (x_diff + x_inc - 2 * theta_m2)
+    lambda_ = (np.sin(np.radians(alpha)) + np.sin(np.radians(beta))) / (m * gratings[grating]["a0"])
+    e_ph = (12398.4197 / lambda_) * 1e-7  # energy in eV
+
+    return e_ph
+
+
+# def _cff(e_ph, grating, r2, r1=_r1, m=_m, gratings=_gratings):
```
Comment (awalter-bnl) [inline startup/31-pgm-energy.py]: Should we remove this line?

File: startup/31-pgm-energy.py [hunk 2]
```diff
@@ -0,0 +1,472 @@
...
+
+
+# def _get_pgm_angles(e_ph, grating, r2, r1=_r1, m=_m, gratings=_gratings,
```
Comment (awalter-bnl) [inline startup/31-pgm-energy.py]: Another line we can probably remove?

File: startup/31-pgm-energy.py [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+        self._readback = self._value  # needs to be in deg too.
+
+    def set(self, value):
+        super().set(np.radians(value - 90.0) * 1e3)  # updates Sirepo model, needs to be in mrad.
+        self.parent.grating_angle._readback = float(value)  # in degrees.
+        self._value = float(value)
+        return NullStatus()
+
+
+# class GrooveDensitySignal(SirepoSignalWithParent):
```
Comment (awalter-bnl) [inline startup/31-pgm-energy.py]: We should probably remove this class which we aren't using.

File: startup/31-pgm-energy.py [hunk 4]
```diff
@@ -0,0 +1,472 @@
...
+            self.parent.pre_mirror_angle.get(),
+            self.parent.grating_angle.get(),
+            m=_m,
+            grating=value,
+            gratings=_gratings,
+            x_inc=_x_inc,
+            x_diff=_x_diff,
+            b=_b,
+        )
+        # print(f"{energy = }")
```
Comment (awalter-bnl) [inline startup/31-pgm-energy.py]: Another Line that probably doesn't need to still be here.

File: startup/31-pgm-energy.py [hunk 5]
```diff
@@ -0,0 +1,472 @@
...
+
+        for k, v in _gratings[value].items():
+            # print(k, v)
```
Comment (awalter-bnl) [inline startup/31-pgm-energy.py]: ANother Print line to remove!-).

File: startup/31-pgm-energy.py [hunk 6]
```diff
@@ -0,0 +1,472 @@
...
+        self.parent.energy._value = energy
+        self.parent.cff.get()
+
+    def set(self, *args, **kwargs):
+        self.put(*args, **kwargs)
+        return NullStatus()
+
+
+class PGMEnergySignal(SignalWithParent):
+    # def __init__(self, value, *args, **kwargs) -> None:
```
Comment (awalter-bnl) [inline startup/31-pgm-energy.py]: Again commented-put lines that should probably be removed

File: startup/31-pgm-energy.py [hunk 7]
```diff
@@ -0,0 +1,472 @@
...
+
+        _r2 = self.parent._r2.get()
+        _r1 = self.parent._r1.get()
+        _m = self.parent._m.get()
+        _b = self.parent._b.get()
+        _x_inc = self.parent._x_inc.get()
+        _x_diff = self.parent._x_diff.get()
+        _cff = self.parent.cff.get()
+
+        # e_ph, grating, r2, r1, m, gratings, x_inc, x_diff, b
```
Comment (awalter-bnl) [inline startup/31-pgm-energy.py]: Again consider removing line

File: startup/31-pgm-energy.py [hunk 8]
```diff
@@ -0,0 +1,472 @@
...
+            cff=_cff,
+        )
+
+        self.parent.pre_mirror_angle.set(theta_pm)
+        self.parent.grating_angle.set(theta_gr)
+
+        return NullStatus()
+
+
+# # define the instantiation parameters here.
```
Comment (awalter-bnl) [inline startup/31-pgm-energy.py]: A bunch of commented out lines to be removed.

File: startup/31-pgm-energy.py [hunk 9]
```diff
@@ -0,0 +1,472 @@
...
+        sirepo_param="grooveDensity2",
+    )
+    _a3 = Cpt(
+        SirepoSignalWithParent,
+        value=objects["grating"].grooveDensity3.get(),
+        sirepo_dict=objects["grating"].grooveDensity3._sirepo_dict,
+        sirepo_param="grooveDensity3",
+    )
+
+    # We explicitly remove these components from the Sirepo class to avoid
```
Comment (awalter-bnl) [inline startup/31-pgm-energy.py]: We explicitly removed these so we should delete them.

File: startup/31-pgm-energy.py [hunk 10]
```diff
@@ -0,0 +1,472 @@
...
+    #                                        **self._interp_kwargs)
+    #     return float(interp_func(energy))
+
+
+pgm = PGM(name="pgm")
+pgm.grating_name.set("HighR")
+
+pgm.energy.set(connection.data["models"]["simulation"]["photonEnergy"])
+pgm.kind = "hinted"
+# pgm.energy.kind = "hinted"
```
Comment (awalter-bnl) [inline startup/31-pgm-energy.py]: Final line that proabaly needs to be deleted.

File: startup/30-epu-energy.py [hunk 11]
```diff
@@ -85,5 +90,5 @@ def _get_magn_field(self, energy):
 # plt.scatter(df_harm["magn_field"]-0.05, f(df_harm["magn_field"]-0.05))
```
Comment (awalter-bnl) [inline startup/30-epu-energy.py:112]: Consider removing this commented out line.

**Review by awalter-bnl — COMMENTED** [review]
Hey Maksim, This looks good in general. There are a number of commented out lines (indicated by comments) that I think we should consider deleting prior to merging.

**Comment by mrakitin** [comment]
Angles before and after setting the energy to 400 eV:
```py

Before
             ('pgm_pre_mirror_angle',
              {'value': 91.94403078738465, 'timestamp': 1685115496.573704}),
             ('pgm_grating_angle',
              {'value': 91.42050527393846, 'timestamp': 1685115496.57388}),

After
             ('pgm_pre_mirror_angle',
              {'value': 91.92603408386577, 'timestamp': 1685115496.573704}),
             ('pgm_grating_angle',
              {'value': 92.45444963631378, 'timestamp': 1685115496.57388}),

```

In [mrad]:
```py
In [17]: np.radians(91.94403078738465 - 90) * 1e3
Out[17]: 33.92973799999995

In [18]: np.radians(91.42050527393846 - 90) * 1e3
Out[18]: 24.792494072170122

In [20]: np.radians(91.92603408386577 - 90) * 1e3
Out[20]: 33.615636269090324

In [21]: np.radians(92.45444963631378 - 90) * 1e3
Out[21]: 42.83822747805288
```

**Comment by mrakitin** [comment]
Ready for another round of review/merge.

**Comment by mrakitin** [comment]
References to the relevant Sirepo issues regarding mirrors in Shadow simulations:
- https://github.com/radiasoft/sirepo/issues/5897
- https://github.com/radiasoft/sirepo/issues/5898

**Review by awalter-bnl — APPROVED** [review]
Looks ready for merging from my side, I note that their is an open review request for @AbbyGi so I will await her review prior to merging unless @mrakitin wants me to merge now !-).

**Review by AbbyGi — APPROVED** [review]

**Comment by mrakitin** [comment]
Thanks, @AbbyGi!
@awalter-bnl, feel free to merge :)

**Pull request merged**